### PR TITLE
Create decimal drop down

### LINKED ISSUE #148: Do not require users to choose a data type for Decimals
- [ ] Optional Decimal fields render as number, null dropdown
 - [ ] Decimal fields render as number
 - [ ] When a user saves the json file, the number fields are saved as strings. For example `"weight": "3.25"`
 - [ ] When a user loads a json file with a string input, it will fill in the field correctly

### PR DESCRIPTION
**Pull request by micahwoodard**
fixes #148 

- removes string and number from options
- adds decimal option
- add regex so only decimals are allowed

### CODE REVIEW

**Comment by micahwoodard** [comment]
Hi Helen I've changed the schemaHandler so it deals with decimals. It saves as a string now and also complains if you have a letters in the text box. The error message isn't super descriptive though and also I haven't made any unit tests for this. Let me know if you think this is the right approach or you were thinking of something else!

**Comment by micahwoodard** [comment]
I've also been trying to create a custom textbox widget that will only allow you to input numeric and decimal values. Still working on the right regular expression though

**Comment by micahwoodard** [comment]
Okay regular expression fixed so only numerical or decimal values allowed

**Review by jtyoung84 — COMMENTED** [review]
I think there might situations where a user needs to input a negative number. Also, we should remove the restriction on the number of digits after the decimal point.

**Review by jtyoung84 — APPROVED** [review]

File: src/custom-ui/CustomWidgets.js [hunk 1]
```diff
@@ -66,6 +66,15 @@ const CustomTextWidget = (props) => {
     readonly={true}
     value={schema.const}
   />
+  } else if (schema.title === 'decimal') {
+    const onChange = (val) => {
+      if (val === undefined || val.match(/^-?\d{0,}(\.\d{0,})?$/)) {
+        props.onChange(val)
+      }
+    }
+    // props.onChange(formattedTime)
```
Comment (helen-m-lin) [inline src/custom-ui/CustomWidgets.js:75]: Remove comment
```suggestion
```

File: src/custom-ui/CustomWidgets.js [hunk 2]
```diff
@@ -66,6 +66,15 @@ const CustomTextWidget = (props) => {
     readonly={true}
     value={schema.const}
   />
+  } else if (schema.title === 'decimal') {
+    const onChange = (val) => {
+      if (val === undefined || val.match(/^-?\d{0,}(\.\d{0,})?$/)) {
```
Comment (helen-m-lin) [inline src/custom-ui/CustomWidgets.js:71]: This regex does not match the one used in schemaHandler.js
```suggestion
      if (val === undefined || val.match(/^-?\d{1,}(\.\d{0,})?$/)) {
```

File: src/custom-ui/CustomWidgets.test.js [hunk 3]
```diff
@@ -35,4 +35,45 @@ describe('CustomTextWidget', () => {
...
+    fireEvent.change(textbox, {target: {value: 'abc'}})
+    expect(textbox.value).toBe('')
+
+    fireEvent.change(textbox, {target: {value: '45abc'}})
+    expect(textbox.value).toBe('')
+
+    fireEvent.change(textbox, {target: {value: '45'}})
+    fireEvent.change(textbox, {target: {value: '45abc'}})
+    expect(textbox.value).toBe('45')
+  })
```
Comment (helen-m-lin) [inline src/custom-ui/CustomWidgets.test.js:77]: If you have time, it would be great to also add a quick test for a negative decimal and a longer decimal!

File: src/utilities/schemaHandler.test.js [hunk 4]
```diff
@@ -196,3 +235,13 @@ test('Checks preProcessSchema does not modify simple sample schema', () => {
   const processedSchema = preProcessSchema(testSchema4)
   expect(processedSchema).toEqual(testSchema4)
 })
+
+test('Checks if anyOf string/number are changed to decimal', () => {
+  const processedSchema = preProcessSchema(testSchema5)
+  expect(processedSchema.properties.test.anyOf).toStrictEqual([{ title: 'decimal', type: 'string', pattern: '^\\d+(\\.\\d{1,2})?$' }])
```
Comment (helen-m-lin) [inline src/utilities/schemaHandler.test.js:241]: Tests are failing because this needs to be updated to be the same as the one in schemaHandlers.js:
```suggestion
  expect(processedSchema.properties.test.anyOf).toStrictEqual([{ title: 'decimal', type: 'string', pattern: '^-?\\d+(\\.\\d{1,})?$' }])
```

File: src/utilities/schemaHandler.test.js [hunk 5]
```diff
@@ -196,3 +235,13 @@ test('Checks preProcessSchema does not modify simple sample schema', () => {
...
+})
+
+test('Checks if anyOf null/string/number are changed to decimal', () => {
+  const processedSchema = preProcessSchema(testSchema6)
+  expect(processedSchema.properties.test.anyOf).toStrictEqual([{ type: 'null', title: 'null' }, { title: 'decimal', type: 'string', pattern: '^\\d+(\\.\\d{1,2})?$' }])
```
Comment (helen-m-lin) [inline src/utilities/schemaHandler.test.js:246]: Tests are failing because this needs to be updated to be the same as the one in schemaHandlers.js:
```suggestion
  expect(processedSchema.properties.test.anyOf).toStrictEqual([{ type: 'null', title: 'null' }, { title: 'decimal', type: 'string', pattern: '^-?\\d+(\\.\\d{1,})?$' }])
```

File: src/utilities/schemaHandlers.js [hunk 6]
```diff
@@ -38,14 +38,26 @@ const preProcessHelper = (obj) => {
...
       if (prop.anyOf) {
-        Object.values(prop.anyOf).forEach(option => {
-          // if the allowed type/ subschema is not a ref nor has a title,
-          // it defaults to <parentProp.title> option 1, <prop.title> option 2, ...
-          // we can convert to display the type name instead
-          if (!option.$ref && option.type && !option.title) {
-            option.title = option.type
-          }
-        })
+        const options = Object.values(prop.anyOf)
```
Comment (helen-m-lin) [inline src/utilities/schemaHandlers.js:41]: nit: we can optimize this a little to avoid doing multiple joins, example:

```js
const options = Object.values(prop.anyOf).map(option => option.type).sort().join()
if (options === 'number,string') {
  prop.anyOf = [{ title: 'decimal', type: 'string', pattern: '^-?\\d+(\\.\\d{1,})?$' }]
} else if (options === 'null,number,string') {
  prop.anyOf = [{ type: 'null', title: 'null' }, { title: 'decimal', type: 'string', pattern: '^-?\\d+(\\.\\d{1,})?$' }]
} else {
  ...
```

**Review by helen-m-lin — CHANGES_REQUESTED** [review]
Thanks Micah! Looks good, but some of the regexes are out of sync with each other. I think eventually we can move them to the Config/constants module so they aren't hardcoded in multiple places, but we can do that at a later point.

Do you mind also running `npm run lint:fix`? This will fix the styling errors. We haven't set up github actions to run this on PRs automatically yet.

File: src/custom-ui/CustomWidgets.js (hunk 2 above)
Comment (micahwoodard) [reply src/custom-ui/CustomWidgets.js:71]: Yes it needs to not match in order top allow the user to type in a value. So this regex allows a user to type just '-' as a start to a number but the validator will complain if you try and submit just '-'. Same with the decimal place. User needs to be able to type '2.'  as a start but '2.' will not be valid.

File: src/custom-ui/CustomWidgets.test.js (hunk 3 above)
Comment (micahwoodard) [reply src/custom-ui/CustomWidgets.test.js:77]: Oh yes will do that makes sense

File: src/custom-ui/CustomWidgets.js (hunk 2 above)
Comment (helen-m-lin) [reply src/custom-ui/CustomWidgets.js:71]: I see, that makes sense! Then the ".7" test will need to be updated to allow ".7". It is failing now.

**Review by helen-m-lin — APPROVED** [review]
